### ISSUE TITLE
Android app [playstore/apk] refuses to start until it can connect to raw.githubusercontent.com

### ISSUE THREAD

**Issue by working-name** [BlueBubblesApp/bluebubbles-app#2691]
Hello!

It seems there's an extremely generous timeout or something else going on but if the app can't access raw.githubusercontent.com, it'll just display a black screen of death.

I had the latest PlayStore version as of 3/1/24 and I thought it croaked. I force quit it multiple times, erased cache. Finally erased data. Uninstalled the app and then installed 1.13.0 from github releases (stable). Same thing, black screen on start, until I allowed it to connect.

Can we make that optional? Updates check, whatever, but let's not stop the user from using the app at all if they don't have internet.

**Comment by SpaceSaver** [comment]
Is this if you attempt to open after installing, but haven't gone through setup?

**Comment by working-name** [comment]
No, regular runs too.

**Comment by SpaceSaver** [comment]
For me, I can force kill and open when mobile data and wifi are off with no issue.

**Comment by working-name** [comment]
Might need to ignore the tablet for a couple weeks then try to start it up again. 

Point is, app will stay black waiting to resolve or connect which is not ideal.

**Comment by SpaceSaver** [comment]
Would you mind sending a screen recording of opening the app?

**Comment by working-name** [comment]
Not at all. Next time it happens.

**Comment by SpaceSaver** [comment]
When does it happen?

**Comment by working-name** [comment]
Whenever the app decides it's time to connect to `raw.githubusercontent.com` on startup. I have a tablet I don't use often and it seems to happen after a certain period of time.

Things I've tried:
- reboot tablet
- force quit app
- empty cache
- empty data (this would prompt a new setup)
- uninstall playstore app and install github APK

None of the above made it get past the black screen of death until I allowed it to connect.

**Comment by zlshames** [comment]
prioritizing this as the update checker should not be invoked if the app is installed via the play store. Moreover, it should not block the main thread

**Comment by tneotia** [comment]
Its definitely not the update checker, I can't even reproduce the issue. Not seeing any calls to that URL in the network inspector. The update checker is guarded behind try/catch and runs async after the app has connected to the server

**Comment by zlshames** [comment]
@working-name do you have any additional details about the URL that's being hit. Looking at the code, we never hit that domain manually. And I don't believe there are any packages that will hit that either.

**Comment by working-name** [comment]
@zlshames 

I'm sorry, I don't have a way to see the full URL without a MITM or something. I also don't know if Android is as straightforward as iOS in terms of system-wide proxy setup.

Anyway, I erased the cache and data, removed all allow rules in ReThink DNS and started the app after force-quitting it. It's stuck on a black screen with the logo after having attempted to connect to `raw.githubusercontent.com` once. That's what lead me to believe that it's a very generous connect timeout.

![photo_2024-04-17_20-28-52](https://github.com/BlueBubblesApp/bluebubbles-app/assets/16505052/09231b67-90e4-468c-8fa7-ec22514624b7)

**Comment by zlshames** [comment]
> @zlshames 
> 
> I'm sorry, I don't have a way to see the full URL without a MITM or something. I also don't know if Android is as straightforward as iOS in terms of system-wide proxy setup.
> 
> Anyway, I erased the cache and data, removed all allow rules in ReThink DNS and started the app after force-quitting it. It's stuck on a black screen with the logo after having attempted to connect to `raw.githubusercontent.com` once. That's what lead me to believe that it's a very generous connect timeout.
> 
> ![photo_2024-04-17_20-28-52](https://github.com/BlueBubblesApp/bluebubbles-app/assets/16505052/09231b67-90e4-468c-8fa7-ec22514624b7)
> 

I installed Rethink DNS and checked the network requests that the app was making, and I did not see the same results as you, unfortunately. To be quite honest, I'm not sure where that request would be coming from. 

The only requests I see are to firebase and my cloudflare domain. I know you are seeing that request, but I just have no idea where it would be coming from.

**Comment by working-name** [comment]
I believe this app running now is from github. But as I stated in my initial post, the behavior is reproducible with the Play Store version of the app too.

**Comment by working-name** [comment]
I can't intercept past the `CONNECT` request because of certificate pinning, don't have a rooted device, nor the time to repack your APK with the modified certificate settings for now. Will try this later when I get some more time. 

Does your app use `Go-http-client/1.1` as the underlying utility? That's its UserAgent in that request.

I am quite sure it's Blue Bubbles making this request given the following:

1. Every time I restart BB a new single request is made to the aforementioned URL.
2. If I allow the connection, BB fully loads.

**Comment by working-name** [comment]
> To be quite honest, I'm not sure where that request would be coming from.

What about the welcome screen? I see some gstatic.com requests too (google fonts?).

**Comment by zlshames** [comment]
> I can't intercept past the `CONNECT` request because of certificate pinning, don't have a rooted device, nor the time to repack your APK with the modified certificate settings for now. Will try this later when I get some more time. 
> 
> Does your app use `Go-http-client/1.1` as the underlying utility? That's its UserAgent in that request.
> 
> I am quite sure it's Blue Bubbles making this request given the following:
> 
> 1. Every time I restart BB a new single request is made to the aforementioned URL.
> 2. If I allow the connection, BB fully loads.

Ok, I was able to reproduce it. @tneotia it only happens when opening to the setup flow, so on a fresh installation, before the sync. So that's good to know. 

We will look into it further. Definitely odd that the request only happens on first start. I'm almost wondering if it's a flutter thing. Either way, at least we have a starting point now

**Comment by zlshames** [comment]
@working-name @tneotia ok, im fairly certain i found the piece of code making that request. It's the network_tools library, specifically this block of code: https://github.com/osociety/network_tools/blob/dev/lib/src/device_info/vendor_table.dart#L30-L46

Which makes a request to: https://raw.githubusercontent.com/osociety/network_tools/main/lib/assets/mac-vendors-export.csv

This library, and this API call allows the package to identify a device based on it's MAC Address prefix. It will _only call the API (csv file) if the file doesn't already exist_. Which is why we weren't seeing the request happen on already setup instances of the app. My hunch aligns with what you were saying in that the API request gets hung up, and doesn't allow the app to proceed. If the API request _wasn't_ hung up, it would have written the file, and thus continued on the next app-open.

That said, there is not an explicit way to disable this call via the library. The mapping is used when instantiating an `ActiveHost` object, which is used for port scanning, which we use for the "detect localhost" feature. Though, we do not use the `.vendor` value in any of our code, so it's not technically required.

We could fork the repository or submit a pull request to it so they can update it for us. But that's the only thing I can think of.

@working-name is there a reason the request is not going through? Is it being blocked?

**Comment by working-name** [comment]
> @working-name is there a reason the request is not going through? Is it being blocked?

No particular reason I just try to isolate app traffic and only allow what I consider acceptable traffic. That being said I just thought BB was looking for updates. I was expecting a failing connection to be gracefully handled - which it sounds like you're already doing given @tneotia's review. So I think you're all good, thank you!

I might raise an issue with the library for a graceful failure of that download attempt.

TL;DR
Trying to control as much of the data hemorrhaging as I can across all my devices, been doing this for years. Currently: Little Snitch on macOS, Portmaster on Windows (although very buggy and very poorly thought-through interface), Rethink DNS on android, AdGuard on iOS with custom filters.

**Comment by zlshames** [comment]
> > @working-name is there a reason the request is not going through? Is it being blocked?
> 
> No particular reason I just try to isolate app traffic and only allow what I consider acceptable traffic. That being said I just thought BB was looking for updates. I was expecting a failing connection to be gracefully handled - which it sounds like you're already doing given @tneotia's review. So I think you're all good, thank you!
> 
> I might raise an issue with the library for a graceful failure of that download attempt.
> 
> TL;DR Trying to control as much of the data hemorrhaging as I can across all my devices, been doing this for years. Currently: Little Snitch on macOS, Portmaster on Windows (although very buggy and very poorly thought-through interface), Rethink DNS on android, AdGuard on iOS with custom filters.

Our app does not gracefully handle a connection failure at that step. It'll fail and prevent the execution of the app, which is what you are experiencing. Don't worry, I've moved the call so that the app startup doesn't depend on it. Will close this after our release
